### PR TITLE
kubectl: add rollout pause/resume subcommands

### PR DESCRIPTION
**Pull request by 0xmichalis**
@ironcladlou @bgrant0607 @madhusudancs @nikhiljindal @janetkuo @kubernetes/kubectl

### CODE REVIEW

File: pkg/kubectl/cmd/rollout/rollout_pause.go [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+		FilenameParam(enforceNamespace, o.Filenames...).
+		ResourceTypeOrNameArgs(true, args...).
+		SingleResourceType().
+		Latest().
+		Do().Infos()
+	if err != nil {
+		return err
+	}
+	if len(infos) != 1 {
+		return fmt.Errorf("rollout pause is only supported on individual resources and resource collections - %d resources were found", len(infos))
```
Comment (bgrant0607) [inline pkg/kubectl/cmd/rollout/rollout_pause.go]: What does "resource collection" mean? Could I apply this to, say, all the deployments matching a label selector?

File: pkg/kubectl/cmd/util/factory.go [hunk 2]
```diff
@@ -281,6 +285,50 @@ func NewFactory(optionalClientConfig clientcmd.ClientConfig) *Factory {
...
+				return false, err
+			}
+
+			switch t := object.(type) {
+			case *extensions.Deployment:
+				if t.Spec.Paused {
+					return true, nil
+				}
+				t.Spec.Paused = true
+				_, err := c.Extensions().Deployments(t.Namespace).Update(t)
```
Comment (bgrant0607) [inline pkg/kubectl/cmd/util/factory.go]: Maybe not a big deal, but we could patch instead.

File: pkg/kubectl/cmd/rollout/rollout_pause.go (hunk 1 above)
Comment (kargakis) [inline pkg/kubectl/cmd/rollout/rollout_pause.go]: Hm, this a copy/paste artifact that I didn't fix. We only allow one resource for now (`len(infos) != 1`). We can enable bulk pausing/resuming in a future iteration.

Comment (kargakis) [inline pkg/kubectl/cmd/rollout/rollout_pause.go]: updated

File: pkg/kubectl/cmd/util/factory.go (hunk 2 above)
Comment (kargakis) [inline pkg/kubectl/cmd/util/factory.go]: I will look into switching to patch in a follow-up if that's ok

Comment (bgrant0607) [inline pkg/kubectl/cmd/util/factory.go]: That's fine.

**Comment by bgrant0607** [comment]
Please file issues to support multiple resources and to use patch instead of update.

LGTM, but not adding label until #19882 merges.

**Comment by kargakis** [comment]
```
â€¢ Failure [17.811 seconds]
Port forwarding
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/portforward.go:240
  With a server that expects a client request
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/portforward.go:203
    should support a client that connects, sends no data, and disconnects [Conformance] [It]
    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/portforward.go:156

    Jan 31 14:48:27.921: Missing "Accepted client connection" from log: 

    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/portforward.go:250
```
unrelated flake

**Comment by kargakis** [comment]
@k8s-bot unit test this

**Comment by kargakis** [comment]
@bgrant0607 rollout is now in master, can you label this appropriately? Thanks

**Comment by kargakis** [comment]
Spawned https://github.com/kubernetes/kubernetes/issues/20435 and https://github.com/kubernetes/kubernetes/issues/20437

**Comment by bgrant0607** [comment]
LGTM label added.

**Comment by bgrant0607** [comment]
And CLA

**Pull request merged**